docs/concepts.md

```diff
@@ -0,0 +1,320 @@
+
+In Skenario the main toplevel Models are the KnativeAutoscaler and the Cluster.
+These establish the various initial Stocks (eg. RequestsBuffered, ReplicasLaunching)
+that are used during the life of the simulation.
+
+Generally speaking, keep logic out of Models. Everything that can be expressed
+as a Movement ought to be; special-case logic should be placed in `Add()` or
+`Remove()`.
+
+### Environment
+
+The Environment is the effective root of the program. Its `Run()` method has the
+core simulation loop and the `AddToSchedule()` method provides the means for
+simulation components to set up future Movements. The Environment also provides a
+number of accessors to key global variables: the simulation's `CurrentMovementTime()`,
+the simulation's `HaltTime()` and a `Context()` that holds a logger object as a value.
+
+The Environment schedules two specialised Movements: the `start_to_running` Movement
+and the `running_to_halted` Movement. These are placed at the boundary points of time.
+That is: only the start Movement may occur at time zero, only the halt Movement may
+occur at the halting time. Other Movements must occur between these two, meaning that
+the permissible time range is expressible as `(0, halt)` or as `[1, halt-1]`.
+
+## How a scenario is executed
+
+### `Run()`
+
+The Environment's `Run()` method implements the "next-event list" concept from
+Discrete Event Simulation. This means that on each iteration, the Environment will
+select the next Movement from a queue and execute it.
+
+Ordering is by the `OccursAt` time of Movements. Internally, the Environment is
+relying on a `MovementPriorityQueue` to maintain orderly records; ultimately this is
+a basic heap.
+
+Once a Movement has been dequeued, the simulation's current time is advanced to
+the `OccursAt` value of the Movement. The Environment then calls the `Remove()` method
+of the `From()` Stock to retrieve an Entity. If that call is successful, it then
+`Add()`s that Entity to the `To()` stock.
+
+On each iteration, Movements that occurred without error are captured in the
+`CompletedMovements` array for later display. Movements that encountered an error are
+stored in the `IgnoredMovements` instead.
+
+### `AddToSchedule()`
+
+This method is how new Movements are scheduled for simulation. Any object with a
+reference to the Environment may call this method. It is the sole mechanism for
+inserting elements into the list.
+
+The Environment will only accept Movements which will occur during the remaining life
+of the simulation. This means it will reject Movements scheduled before the current
+time and it will reject events that would occur after the halt time. Such Movements
+are added to the `IgnoredMovements` array.
+
+The Environment also rejects new Movements when an existing Movement is already
+scheduled to `OccurAt` that time. This is partly due to the underlying implementation,
+which requires a strict ordering of entries in the heap. But it also reflects that
```

Comment:
Presumably strict ordering is necessary so that the simulator is deterministic? Otherwise, you could just arbitrarily choose between Movements which occur at the same time.